include/unifex/get_stop_token.hpp

```diff
@@ -16,15 +16,41 @@
 #pragma once
 
 #include <unifex/stop_token_concepts.hpp>
+#include <unifex/inplace_stop_token.hpp>
 #include <unifex/tag_invoke.hpp>
 #include <unifex/type_traits.hpp>
+#include <unifex/coroutine.hpp>
 #include <unifex/unstoppable_token.hpp>
 
 #include <unifex/detail/prologue.hpp>
 
 namespace unifex {
 namespace _get_stop_token {
   inline const struct _fn {
+  private:
+#if !UNIFEX_NO_COROUTINES
+    struct awaiter_ {
+      // This will fail for coroutine promises storing other kinds of
+      // stop tokens, but we don't currently have examples of that right
+      // now.
+      inplace_stop_token stoken_;
+      bool await_ready() const noexcept {
+        return false;
+      }
+      template <typename Promise>
+      bool await_suspend(coro::coroutine_handle<Promise> coro) noexcept {
+        stoken_ = _fn{}(coro.promise());
+        return false; // don't suspend
+      }
+      inplace_stop_token await_resume() const noexcept {
+        return stoken_;
+      }
+    };
+    friend awaiter_ operator co_await(_fn) noexcept {
+      return {};
+    }
+#endif
```

Comment:
I think in time this should be generalised to allow the `get_stop_token` CPO to be used as a sender that will just return the result of calling `get_stop_token()` on the receiver it is connected to.
This could be made a general pattern which could be encapsulated, either in a base-class or a higher-order algorithm.
e.g. `query(get_stop_token)` could be a syntax for producing such a sender.

However, I think for this to work in coroutines it needs to be a `typed_sender` which currently doesn't let the return-type depend on the receiver type. For that we need to move the type-traits for senders into the operation-state so that will have to wait.

However, this implementation assumes that the only coroutine that might await this will have a stop-token type of `inplace_stop_token` which is not necessarily always going to be the case.

This would be better implemented by customising the `await_transform()` CPO, which provides the promise object/type, and then returning an awaiter whose `await_resume()` method returns the appropriate stop-token type deduced from calling the CPO on the promise.

Reply:
Also, I'm not sure about just having the CPO itself be directly awaitable.

What was the thinking between having `co_await get_stop_token;` vs `co_await get_stop_token();`?

Reply:
> What was the thinking between having `co_await get_stop_token;` vs `co_await get_stop_token();`?

I don't care which. Seems like you would prefer `co_await get_stop_token()`, is that right?